monix-reactive/shared/src/main/scala/monix/reactive/internal/operators/GroupByOperator.scala

```diff
@@ -51,7 +51,7 @@ private[reactive] final class GroupByOperator[A, K](
       private def retryOnNext(elem: A): Future[Ack] =
         onNext(elem)
 
-      @tailrec def onNext(elem: A): Future[Ack] =
```

Comment:
Are you sure you want to remove all those `@tailrec` annotations? We could rewrite the functions to make them tail recursive.

Reply:
They are still tail recursive, it just can't be verified with the annotation, right?

It would be preferable to keep the annotation but I'm not sure how to do that in cases like:

```scala
  def unsafeSubscribeFn(subscriber: Subscriber[A]): Cancelable = {
    import subscriber.scheduler
    val state = stateRef.get()

    if (state.errorThrown != null) {
      subscriber.onError(state.errorThrown)
      Cancelable.empty
    } else if (state.isDone) {
      Observable
        .now(state.cached)
        .unsafeSubscribeFn(subscriber)
    } else {
      val c = ConnectableSubscriber(subscriber)
      val newState = state.addNewSubscriber(c)

      if (stateRef.compareAndSet(state, newState)) {
        c.pushFirst(state.cached)
        val connecting = c.connect()

        val cancelable = Cancelable { () =>
          removeSubscriber(c)
        }
        connecting.syncOnStopOrFailure(_ => cancelable.cancel())
        cancelable
      } else {
        // retry
        unsafeSubscribeFn(subscriber)
      }
    }
  }
```

Reply:
It's true, but in case something changes in the future, the compiler won't tell us if it's still tail recursive. I refactored this in #1326 by extracting the core loop into an inner method that I've annotated accordingly. However, I don't have any strong opinions on this.

Reply:
Ahh, I was too hasty - I've corrected these method in the latest commit